qos/cometbft/context.go

```diff
@@ -0,0 +1,164 @@
+
+// requestContext provides the support required by the gateway
+// package for handling service requests.
+var _ gateway.RequestQoSContext = &requestContext{}
+
+type response interface {
+	GetObservation() qosobservations.CometBFTEndpointObservation
+	GetResponsePayload() []byte
+	GetResponseStatusCode() int
+}
+
+type endpointResponse struct {
+	protocol.EndpointAddr
+	response
+	unmarshalErr error
+}
+
+// requestContext provides the functionality required
+// to support QoS for a CometBFT blockchain service.
+type requestContext struct {
+	logger polylog.Logger
+
+	httpReq       *http.Request
+	endpointStore *EndpointStore
+
+	// isValid indicates whether the underlying user request
+	// for this request context was found to be valid.
+	// This field is set by the corresponding QoS instance
+	// when creating this request context during the parsing
+	// of the user request.
+	isValid bool
+
+	// preSelectedEndpointAddr allows overriding the default
+	// endpoint selector with a specific endpoint's addresss.
+	// This is used when building a request context as a check
+	// for a specific endpoint.
+	preSelectedEndpointAddr protocol.EndpointAddr
+
+	// endpointResponses is the set of responses received from one or
+	// more endpoints as part of handling this service request.
+	// NOTE: these are all related to a single JSONRPC request,
+	// enhancing to support batch JSONRPC requests will involve the
+	// modification of this field's type.
+	endpointResponses []endpointResponse
+}
+
+func (rc requestContext) GetServicePayload() protocol.Payload {
+	return protocol.Payload{
+		//CometBFT uses only GET HTTP requests with empty body
+		Data: "",
+
+		// Method is alway GET for CometBFT-based blockchains.
+		Method: http.MethodGet,
+
+		// CometBFT uses REST-like routes, e.g. `/status`.
+		Path: rc.httpReq.URL.Path,
+
+		// TODO_IMPROVE: adjust the timeout based on the request method:
+		// An endpoint may need more time to process certain requests,
+		// as indicated by the request's method and/or parameters.
+		TimeoutMillisec: defaultServiceRequestTimeoutMillisec,
+	}
+}
+
+// UpdateWithResponse is NOT safe for concurrent use
+func (rc *requestContext) UpdateWithResponse(endpointAddr protocol.EndpointAddr, responseBz []byte) {
+	// TODO_IMPROVE: check whether the request was valid, and return an error if it was not.
+	// This would be an extra safety measure, as the caller should have checked the returned value
+	// indicating the validity of the request when calling on QoS instance's ParseHTTPRequest
+
+	response, err := unmarshalResponse(rc.logger, rc.httpReq.URL.Path, responseBz)
+
+	rc.endpointResponses = append(rc.endpointResponses,
+		endpointResponse{
+			EndpointAddr: endpointAddr,
+			response:     response,
+			unmarshalErr: err,
+		},
+	)
+}
+
+// GetHTTPResponse builds the HTTP response that should be returned for
+// a CometBFT blockchain service request.
+func (rc requestContext) GetHTTPResponse() gateway.HTTPResponse {
```

Comment:
Can we please add a TODO to consider some refactoring of different components? Many of these play an identical role in each QoS service implementation but with different details.

This "duplication" is by design of course, but leaving a few TODOs re refactoring will help future readers  understand the context more efficiently.

Reply:
> Can we please add a TODO to consider some refactoring of different components? Many of these play an identical role in each QoS service implementation but with different details.
> 
> This "duplication" is by design of course, but leaving a few TODOs re refactoring will help future readers understand the context more efficiently.

Yes I agree 100% as well. I will add TODOs to capture this work for a future PR.